File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    private void rotting(Set<Integer> freshes, Set<Integer> rottens, int index) {
        if (freshes.contains(index)) {
            freshes.remove(index);
            rottens.add(index);
        }
    }

    public int orangesRotting(int[][] grid) {
        Set<Integer> freshes = new HashSet<>();
        Set<Integer> rottens = new HashSet<>();

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                int index = i * grid[i].length + j;
                if (1 == grid[i][j]) {
                    freshes.add(index);
                } else if (2 == grid[i][j]) {
                    rottens.add(index);
                }
            }
        }

        System.out.println(freshes);
        System.out.println(rottens);

        if (freshes.isEmpty()) {
            return 0;
        } 

        int steps = -1;
        while (!rottens.isEmpty()) {
            Set<Integer> newRottens = new HashSet<>();
            for (int index : rottens) {
                int i = index / grid[0].length, j = index % grid[0].length;

                if (0 < j) {
                    rotting(freshes, newRottens, index - 1);
                }

                if (grid[0].length - 1 > j) {
                    rotting(freshes, newRottens, index + 1);
                }

                if (0 < i) {
                    rotting(freshes, newRottens, index - grid[0].length);
                }

                if (grid.length - 1 > i) {
                    rotting(freshes, newRottens, index + grid[0].length);
                }
            }

            steps++;
            rottens = newRottens;
        }

        if (freshes.isEmpty()) {
            return steps;
        }

        return -1;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    # Move from "freshes" into "rottens"
    def rotting(self, freshes: Set(Tuple), rottens: Set(Tuple), i: int, j: int):
        if ((i, j)) in freshes:
            freshes.remove((i, j))
            rottens.add((i, j))

    def orangesRotting(self, grid: List[List[int]]) -> int:
        # Construct initial "freshes" set and "rottens" set
        freshes = set()
        rottens = set()
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if 1 == grid[i][j]:
                    freshes.add((i, j))
                elif 2 == grid[i][j]:
                    rottens.add((i, j))

        # if no fresh at all
        if not freshes:
            return 0

        steps = -1
        while rottens:
            # Initialize the set for those just rotten
            newRottens = set()
            for rotten in rottens:
                i = rotten[0]
                j = rotten[1]
                # Move adjacent fresh(es) into "newRottens"
                self.rotting(freshes, newRottens, i, j - 1)
                self.rotting(freshes, newRottens, i, j + 1)
                self.rotting(freshes, newRottens, i - 1, j)
                self.rotting(freshes, newRottens, i + 1, j)
        
            steps += 1

            # Set "rottens" to "newRottens" for the nexe round
            rottens = newRottens

        # If still any fresh remained
        if freshes:
            return -1

        return steps